affectedmoduledetector/src/main/kotlin/com/dropbox/affectedmoduledetector/ToStringLogger.kt

```diff
@@ -20,52 +20,95 @@
+import org.gradle.api.Project
+import org.gradle.api.file.RegularFileProperty
 import org.gradle.api.logging.LogLevel
 import org.gradle.api.logging.Logger
+import org.gradle.api.provider.Property
+import org.gradle.api.provider.Provider
+import org.gradle.api.services.BuildService
+import org.gradle.api.services.BuildServiceParameters
+import org.gradle.internal.build.event.BuildEventListenerRegistryInternal
 import org.gradle.internal.logging.slf4j.OutputEventListenerBackedLogger
 import org.gradle.internal.logging.slf4j.OutputEventListenerBackedLoggerContext
+import org.gradle.internal.operations.*
 import org.gradle.internal.time.Clock
+import org.gradle.invocation.DefaultGradle
+import java.io.File
 
 /**
  * Gradle logger that logs to a string.
  */
-class ToStringLogger(
-    private val stringBuilder: StringBuilder = StringBuilder()
+internal open class ToStringLogger(
+    private val stringBuilder: Provider<ToStringLoggerBuildService>?
 ) : OutputEventListenerBackedLogger(
         "amd",
-        OutputEventListenerBackedLoggerContext(
-                Clock {
-                    System.currentTimeMillis()
-                }
-        ).also {
+        OutputEventListenerBackedLoggerContext {
+            System.currentTimeMillis()
+        }.also {
             it.level = LogLevel.DEBUG
-            it.setOutputEventListener {
-                stringBuilder.appendln(it.toString())
+            it.setOutputEventListener { outputEvent ->
+                stringBuilder?.get()?.parameters?.getStringBuilderProperty()?.get()?.appendLine(outputEvent.toString())
             }
         },
         Clock {
             System.currentTimeMillis()
         }
 ) {
+
     /**
      * Returns the current log.
      */
-    fun buildString() = stringBuilder.toString()
+    fun buildString() = stringBuilder?.get()?.parameters?.getStringBuilderProperty()?.get()?.toString()
 
+    @Suppress("UnstableApiUsage") // BuildService is not yet stable
     companion object {
+        internal abstract class ToStringLoggerBuildService : BuildService<ToStringLoggerBuildService.ToStringLoggerBuildServiceParameters>, BuildOperationListener, AutoCloseable {
+            interface ToStringLoggerBuildServiceParameters : BuildServiceParameters {
+                fun getStringBuilderProperty(): Property<StringBuilder>
+                fun getOutputFileProperty(): RegularFileProperty
+            }
+
+            override fun started(p0: BuildOperationDescriptor, p1: OperationStartEvent) { }
+
+            override fun progress(p0: OperationIdentifier, p1: OperationProgressEvent) { }
+
+            override fun finished(buildOperationDescriptor: BuildOperationDescriptor, operationFinishEvent: OperationFinishEvent) { }
+
+            override fun close() {
+                val outputFile = parameters.getOutputFileProperty().orNull?.asFile ?: return
+                outputFile.appendText(parameters.getStringBuilderProperty().get().toString())
```

Comment:
frankly speaking, it look quite difficult for understanding and a lot of code. 
Your approach is not regular, hence, would u analyse my variant of this file?

It might be quite important to write easy code in open sourse because sometimes worldwide developers want to modify project for themselvs projects. 

I just used standart approach and it seems easer. But, please, recheck your logic attentevly. 

```kotlin
internal open class ToStringLogger(
    private val loggerProvider: Provider<ToStringLoggerBuildService>?
) : OutputEventListenerBackedLogger(
    "amd",
    OutputEventListenerBackedLoggerContext { System.currentTimeMillis() }
        .apply {
            level = LogLevel.DEBUG
            setOutputEventListener { outputEvent ->
                loggerProvider?.get()?.logBuffer?.appendLine(outputEvent.toString())
            }
        },
    Clock { System.currentTimeMillis() }
) {

    /**
     * Returns the current log.
     */
    fun buildString() = loggerProvider?.get()?.logBuffer?.toString()

    // ATTENTION HERE! -------------------------------------------------------------------------------------------
    @Suppress("UnstableApiUsage") // BuildService is not stable yet
    private interface ToStringLoggerBuildServiceParams : BuildServiceParameters {

        val logBuffer: StringBuilder
        var logFile: File?
    }

    @Suppress("UnstableApiUsage") // BuildService is not stable yet
    internal abstract inner class ToStringLoggerBuildService
        : BuildService<ToStringLoggerBuildServiceParams>,
        BuildOperationListener,
        AutoCloseable,
        ToStringLoggerBuildServiceParams {

        override val logBuffer = StringBuilder()
        override var logFile: File? = null

        override fun started(p0: BuildOperationDescriptor, p1: OperationStartEvent) = Unit
        override fun progress(p0: OperationIdentifier, p1: OperationProgressEvent) = Unit
        override fun finished(
            buildOperationDescriptor: BuildOperationDescriptor,
            operationFinishEvent: OperationFinishEvent
        ) = Unit

        override fun close() {
            val outputFile = logFile ?: return
            outputFile.appendText(logBuffer.toString())
            println("Wrote dependency log to ${outputFile.absolutePath}")
        }
    }

    @Suppress("UnstableApiUsage") // BuildService is not yet stable
    companion object {

        private const val LOGGER_BUILD_LISTENER = "to-string-logger-build-listener"

        /**
         * Creates the [ToStringLogger]
         *
         * @param project the current project to apply to
         * @param logFilename the filename for the logs to go
         * @param logFolder the path to where the log should output. if null doesn't output
         */
        fun createWithLifecycle(
            project: Project,
            logFilename: String,
            logFolder: String? = null
        ): Logger {
            val gradle = project.gradle

            val toStringLoggerBuildService = gradle
                .sharedServices
                .registerIfAbsent(
                    LOGGER_BUILD_LISTENER,
                    ToStringLoggerBuildService::class.java
                ) { service ->
                    if (logFolder != null) {
                        val distDir = File(logFolder)
                        if (!distDir.exists()) {
                            distDir.mkdirs()
                        }
                        val outputFile = distDir.resolve(logFilename)
                        service.parameters.logFile = outputFile
                    }
                }

            val logger = ToStringLogger(toStringLoggerBuildService)
            (gradle as DefaultGradle)
                .services[BuildEventListenerRegistryInternal::class.java]
                .onOperationCompletion(toStringLoggerBuildService)

            return logger
        }
    }
}

```


Best regards :)

Reply:
If you want to fix this code style later please feel free to put up a PR but other then moving the class outside the `companion object` and doing all your other code style changes I don't see much benefit to this alteration